javalib/src/main/scala/java/io/DataOutputStream.scala

```diff
@@ -18,26 +24,57 @@ class DataOutputStream(out: OutputStream)
   }
 
   override def write(b: Int): Unit = {
-    out.write(b)
+    out.write(b) // underlying stream will convert to byte.
     written += 1
   }
 
   override final def writeBoolean(v: Boolean): Unit =
     write(if (v) 1 else 0)
 
   override final def writeByte(v: Int): Unit =
-    write(v.toByte)
+    write(v)
+
+  override final def writeBytes(s: String): Unit = {
+    var index = 0
+
+    for (ch <- s) {
+      if (index == byteArrayCapacity) {
+        write(byteArray, 0, byteArrayCapacity)
+        index = 0
+      }
 
-  override final def writeBytes(s: String): Unit =
-    write(s.toArray.map(_.toByte))
+      byteArray(index) = ch.toByte
+      index += 1
+    }
+
+    if (index > 0)
+      write(byteArray, 0, index)
+  }
 
   override final def writeChar(v: Int): Unit = {
-    write((v >> 8) & 0xFF)
-    write(v & 0xFF)
+    byteArray(0) = (v >> 8).toByte
+    byteArray(1) = (v & 0xFF).toByte
```

Comment:
```suggestion
    byteArray(1) = v.toByte
```
`& 0xFF` is redundant if we're doing `.toByte` after.

Reply:
Done, in several places.